actions/mysql_query_select_action.go

```diff
@@ -0,0 +1,75 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package actions
+
+import (
+	"context"
+
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/pkg/errors"
+)
+
+type mysqlQuerySelectAction struct {
+	id     string
+	params *agentpb.StartActionRequest_MySQLQuerySelectParams
+}
+
+// NewMySQLQuerySelectAction creates MySQL SELECT query Action.
+func NewMySQLQuerySelectAction(id string, params *agentpb.StartActionRequest_MySQLQuerySelectParams) Action {
+	return &mysqlQuerySelectAction{
+		id:     id,
+		params: params,
+	}
+}
+
+// ID returns an Action ID.
+func (a *mysqlQuerySelectAction) ID() string {
+	return a.id
+}
+
+// Type returns an Action type.
+func (a *mysqlQuerySelectAction) Type() string {
+	return "mysql-query-select"
+}
+
+// Run runs an Action and returns output and error.
+func (a *mysqlQuerySelectAction) Run(ctx context.Context) ([]byte, error) {
+	db, err := mysqlOpen(a.params.Dsn)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close() //nolint:errcheck
+
+	// use prepared statement to force binary protocol usage that returns correct types
+	stmt, err := db.PrepareContext(ctx, "SELECT /* pmm-agent */ "+a.params.Query) //nolint:gosec
```

Comment:
We can also add validation that the query doesn't contain a semicolon, to prevent injections. Also probably we can forbid UNION, DROP, TRUNCATE, and other keywords that obviously out of the scope of STT. Maybe it should be done in SAAS repo.

Reply:
You should not be able to write a query like `SELECT … UNION DROP/TRUNCATE …`. But please try it. If one can do that, then we should definitely do something about it.